mal/step8_macros.mal

```diff
@@ -60,111 +52,94 @@
 
     "else"        ast))))
 
-(def! LET (fn* [env args]
-  (if (> (count args) 0)
+(def! LET (fn* [env args form]
+  (if (empty? args)
+    (EVAL form env)
     (do
-      (env-set env (nth args 0) (EVAL (nth args 1) env))
-      (LET env (rest (rest args)))))))
+      (env-set env (first args) (EVAL (nth args 1) env))
+      (LET env (rest (rest args)) form)))))
 
 (def! EVAL (fn* [ast env] (do
-  ;;(do (prn "EVAL" ast "/" (keys @env)) )
-  (if (not (list? ast))
-    (eval-ast ast env)
-
-    ;; apply list
-    (let* [ast (MACROEXPAND ast env)]
-      (if (not (list? ast))
-        (eval-ast ast env)
-
-        (let* [a0 (first ast)]
-          (cond
-            (nil? a0)
-            ast
-
-            (= 'def! a0)
-            (env-set env (nth ast 1) (EVAL (nth ast 2) env))
-
-            (= 'let* a0)
-            (let* [let-env (new-env env)]
-              (do
-                (LET let-env (nth ast 1))
-                (EVAL (nth ast 2) let-env)))
-
-            (= 'quote a0)
-            (nth ast 1)
-
-            (= 'quasiquote a0)
-            (let* [a1 (nth ast 1)]
-              (EVAL (QUASIQUOTE a1) env))
-
-            (= 'defmacro! a0)
-            (let* [a1 (nth ast 1)
-                   a2 (nth ast 2)
-                   f (EVAL a2 env)
-                   m (or (meta f) {})
-                   mac (with-meta f (assoc m "ismacro" true))]
-              (env-set env a1 mac))
-
-            (= 'macroexpand a0)
-            (let* [a1 (nth ast 1)]
-              (MACROEXPAND a1 env))
-
-            (= 'do a0)
-            (let* [el (eval-ast (rest ast) env)]
-              (nth el (- (count el) 1)))
-
-            (= 'if a0)
-            (let* [cond (EVAL (nth ast 1) env)]
-              (if (or (= cond nil) (= cond false))
-                (if (> (count ast) 3)
-                  (EVAL (nth ast 3) env)
-                  nil)
-                (EVAL (nth ast 2) env)))
-
-            (= 'fn* a0)
-            (fn* [& args]
-              (EVAL (nth ast 2) (new-env env (nth ast 1) args)))
-
-            "else"
-            (let* [el (eval-ast ast env)
-                  f (first el)
-                  args (rest el)]
-              (apply f args))))))))))
+  ;; (prn "EVAL" ast "/" (keys @env))
+  (let* [ast (MACROEXPAND ast env)]
+    (if (not (list? ast))
+      (eval-ast ast env)
+
+      ;; apply list
+      (let* [a0 (first ast)]
+        (cond
+          (empty? ast)
+          ast
+
+          (= 'def! a0)
+          (env-set env (nth ast 1) (EVAL (nth ast 2) env))
+
+          (= 'let* a0)
+          (apply LET (new-env env) (rest ast))
+
+          (= 'quote a0)
+          (nth ast 1)
+
+          (= 'quasiquote a0)
+          (QUASIQUOTE (nth ast 1) env)
+
+          (= 'defmacro! a0)
+          (let* [f (EVAL (nth ast 2) env)
+                 m (meta f)]
+            (env-set env (nth ast 1) ^(assoc (if m m {}) "ismacro" true) f))
```

Comment:
I'm okay with a1 and a2 not being pulled out in the `let*` but I think mac still needs to be explicit. This line is doing a lot of heavy lifting and it's hard to parse for me let alone somebody unfamiliar with Lisp. I think `with-meta` is also clearer here than `^`.

Reply:
Last commits should take all your answers into account.

A conflict in `ps/stepA` has been the opportunity to revert the changes you have rejected and avoid contradictions in consecutive commits. It also helped me to realize how much work a simple change like adding `inc` in step A can cause, so I will not insist on adding `inc`, `gensym`, `land` into the step files anymore.

About `quasiquote`: I realize that this issue deserves a separate pull request.

About `macro?`: this function is (re)defined in `core.mal` and the optional built-in is never used. It may be wise to prefix the MAL version `_macro?` as done for `_fn?`.

About `and`
- the name could be `logical-and`, `and2` (as in `lib/equality.mal`) or `&&` (suggesting untyped parameters, lazy evaluation and boolean result to anyone knowing it). Avoiding confusion with Clojure's `and` is easy.
- The diff above shows where `(and ab)` would be more readable than `(if a b)` or `(cond (not a) false :else b)`. The intent of `_fn?` and `macro?` would be more visible without the conversion to boolean. The two `is-pair`, `is-macro-call` functions seem to only exist in order to make a conjunction readable. Do you think the increase in readability is worth an auxiliary `and` function in `core.mal`?
```
--- a/mal/core.mal                                                             
+++ b/mal/core.mal                                                             
@@ -1,14 +1,15 @@
+(defmacro! and (fn* [& xs]                                                    
+  (if (empty? xs)                                                             
+    true                                                                      
+    (list 'if (first xs) (cons 'and (rest xs)) false))))                      
+                                                                              
 (def! _fn? (fn* [x]                                                           
-  (if (fn? x)                                                                 
-    (not (get (meta x) "ismacro"))                                            
-    false)))                                                                  
+  (and (fn? x)                                                                
+       (not (get (meta x) "ismacro")))))                                      
                                                                               
 (def! macro? (fn* [x]                                                         
-  (if (fn? x)                                                                 
-    (if (get (meta x) "ismacro")                                              
-      true                                                                    
-      false)                                                                  
-    false)))                                                                  
+  (and (fn? x)                                                                
+       (get (meta x) "ismacro"))))                                            
                                                                               
 (def! core_ns                                                                 
   [['= =]                                                                     
--- a/mal/stepA_mal.mal                                                        
+++ b/mal/stepA_mal.mal                                                        
@@ -7,8 +7,8 @@
                                                                               
 ;; eval                                                                       
 (def! is-pair (fn* [x]                                                        
-  (if (sequential? x)                                                         
-    (not (empty? x)))))                                                       
+  (and (sequential? x)                                                        
+       (not (empty? x)))))                                                    
                                                                               
 (def! QUASIQUOTE (fn* [ast]                                                   
   (if (not (is-pair ast))
@@ -18,18 +18,21 @@
        (= 'unquote a0)                                                        
        (nth ast 1)                                                            
                                                                               
-       (if (is-pair a0) (= 'splice-unquote (first a0))) ; `if` means `and`    
+       (and (is-pair a0)                                                      
+            (= 'splice-unquote (first a0)))                                   
        (list 'concat (nth a0 1) (QUASIQUOTE (rest ast)))                      
                                                                               
        "else"                                                                 
        (list 'cons (QUASIQUOTE a0) (QUASIQUOTE (rest ast))))))))              
                                                                               
+                                                                              
+                                                                              
 (def! is-macro-call (fn* [ast env]                                            
-  (if (list? ast)                                                             
-    (let* [a0 (first ast)]                                                    
-      (if (symbol? a0)                                                        
-        (if (env-find env a0)                                                 
-          (macro? (env-get env a0))))))))                                     
+  (and (list? ast)                                                            
+       (let* [a0 (first ast)]                                                 
+         (and (symbol? a0)                                                    
+              (env-find env a0)                                               
+              (macro? (env-get env a0)))))))                                  
                                                                               
 (def! MACROEXPAND (fn* [ast env]                                              
   (if (is-macro-call ast env)                                                 
```